runtime/adapters/src/lib.rs

```diff
@@ -789,3 +792,100 @@ where
 		}
 	}
 }
+
+pub struct StableswapHooksAdapter<Runtime>(PhantomData<Runtime>);
+
+impl<Runtime> StableswapHooks<AssetId> for StableswapHooksAdapter<Runtime>
+where
+	Runtime: pallet_ema_oracle::Config + pallet_stableswap::Config,
+{
+	fn on_liquidity_changed(pool_id: AssetId, state: PoolState<AssetId>) -> DispatchResult {
+		let pool_size = state.assets.len();
+
+		// As we access by index, let's ensure correct vec lengths.
+		ensure!(
+			state.before.len() == pool_size,
+			pallet_stableswap::Error::<Runtime>::IncorrectAssets.into()
+		);
+		ensure!(
+			state.after.len() == pool_size,
+			pallet_stableswap::Error::<Runtime>::IncorrectAssets.into()
+		);
+		ensure!(
+			state.delta.len() == pool_size,
+			pallet_stableswap::Error::<Runtime>::IncorrectAssets.into()
+		);
+
+		for idx in 0..pool_size {
+			OnActivityHandler::<Runtime>::on_liquidity_changed(
+				STABLESWAP_SOURCE,
+				state.assets[idx],
+				pool_id,
+				state.delta[idx],
+				0, //TODO: fix
+				state.after[idx],
+				state.shares,
+			)
+			.map_err(|(_, e)| e)?;
+		}
+
+		Ok(())
+	}
+
+	fn on_trade(
+		pool_id: AssetId,
+		_asset_in: AssetId,
+		_asset_out: AssetId,
+		state: PoolState<AssetId>,
+	) -> DispatchResult {
+		let pool_size = state.assets.len();
+
+		// As we access by index, let's ensure correct vec lengths.
+		ensure!(
+			state.before.len() == pool_size,
+			pallet_stableswap::Error::<Runtime>::IncorrectAssets.into()
+		);
+		ensure!(
+			state.after.len() == pool_size,
+			pallet_stableswap::Error::<Runtime>::IncorrectAssets.into()
+		);
+		ensure!(
+			state.delta.len() == pool_size,
+			pallet_stableswap::Error::<Runtime>::IncorrectAssets.into()
+		);
+
+		for idx in 0..pool_size {
+			OnActivityHandler::<Runtime>::on_trade(
+				STABLESWAP_SOURCE,
+				state.assets[idx],
+				pool_id,
+				state.delta[idx],
+				0, // Correct
+				state.after[idx],
+				state.shares,
+			)
+			.map_err(|(_, e)| e)?;
+
+			OnActivityHandler::<Runtime>::on_trade(
+				STABLESWAP_SOURCE,
+				pool_id,
+				state.assets[idx],
+				0, // Correct
+				state.delta[idx],
+				state.shares,
+				state.after[idx],
+			)
+			.map_err(|(_, e)| e)?;
```

Comment:
This looks wrong?
It looks like you trigger the same asset pair twice per trade?

Reply:
removed